### PR TITLE
Fix filling issues more generally

### PR DESCRIPTION
**Pull request by dominikh**
This approach catches more edge cases. It is also easier to understand, but probably not as efficient as it could be.

### CODE REVIEW

**Comment by dominikh** [comment]
I have the following passing test cases now (all fills). Some of these you've
already added yourself, some you haven't.


- a circle at (128., 128.) with radius 64
- a rectangle from (128., 128.) to (160., 160.)

```
path.move_to((4., 0.));
path.line_to((8., 4.));
path.line_to((4., 8.));
path.line_to((0., 4.));
path.close_path();
```
```
path.move_to((4.00001, 1e-45));
path.line_to((8.00001, 4.00001));
path.line_to((4.00001, 8.00001));
path.line_to((1e-45, 4.00001));
path.close_path();
```
```
path.move_to((16., 8.));
path.line_to((31.999998, 8.));
path.line_to((31.999998, 9.));
path.line_to((16., 9.));
path.close_path();
```
```
path.move_to((32.000002, 9.));
path.line_to((28., 9.));
path.line_to((28., 8.));
path.line_to((32.000002, 8.));
path.close_path()
```
```
path.move_to((16., 8.));
path.line_to((16., 9.));
path.line_to((32., 9.));
path.line_to((32., 8.));
path.close_path()
```
```
path.move_to((16.000002, 8.));
path.line_to((20.000002, 8.));
path.line_to((24.000002, 8.));
path.line_to((28.000002, 8.));
path.line_to((32.000002, 8.));
path.line_to((32.000002, 9.));
path.line_to((28.000002, 9.));
path.line_to((24.000002, 9.));
path.line_to((20.000002, 9.));
path.move_to((16.000002, 9.));
path.close_path();
```
```
path.move_to((16.000427, 8.));
path.line_to((20.000427, 8.));
path.line_to((24.000427, 8.));
path.line_to((28.000427, 8.));
path.line_to((32.000427, 8.));
path.line_to((32.000427, 9.));
path.line_to((28.000427, 9.));
path.line_to((24.000427, 9.));
path.line_to((20.000427, 9.));
path.move_to((16.000427, 9.));
path.close_path();
```

**Comment by LaurenzV** [comment]
From what I can tell `issue_2_incorrect_filling_4` still looks wrong?

**Comment by LaurenzV** [comment]
And I wasn't able to reproduce the circle one, what tolerance and canvas size did you use?

**Comment by LaurenzV** [comment]
And performance doesn't matter for now. :) Main goal now is robustness and correctness.

File: tests/issues.rs [hunk 1]
```diff
@@ -36,3 +36,114 @@ fn issue_2_incorrect_filling_2() {
...
+    path.move_to((16.000002, 8.));
+    path.line_to((20.000002, 8.));
+    path.line_to((24.000002, 8.));
+    path.line_to((28.000002, 8.));
+    path.line_to((32.000002, 8.));
+    path.line_to((32.000002, 9.));
+    path.line_to((28.000002, 9.));
+    path.line_to((24.000002, 9.));
+    path.line_to((20.000002, 9.));
+    path.move_to((16.000002, 9.));
```
Comment (dominikh) [inline tests/issues.rs:69]: Mea culpa, this should've been another line_to, not a move_to.

**Comment by dominikh** [comment]
> And I wasn't able to reproduce the circle one, what tolerance and canvas size did you use?

Tolerance 0.1, canvas 256. I can reproduce it on 24d7589. baec1ab already fixed it.

**Comment by LaurenzV** [comment]
> Tolerance 0.1, canvas 256. I can reproduce it on https://github.com/LaurenzV/cpu-sparse-experiments/commit/24d75891a517bc5ae26876c03e7192e36660895e. https://github.com/LaurenzV/cpu-sparse-experiments/commit/baec1abbc6e3db91e0731b98f9992eab67aaf587 already fixed it.

Okay, I will leave it out then since we already have a test case for that. 😄

**Comment by LaurenzV** [comment]
Thanks!
